text/0000-read-all.md

```diff
@@ -0,0 +1,99 @@
+- Feature Name: read_exact and read_full
+- Start Date: 2015-03-15
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+
+Rust's `Write` trait has `write_all`, which is a convenience method that calls
+`write` repeatedly to write an entire buffer. This proposal adds two similar
+convenience methods to the `Read` trait: `read_full` and `read_exact`.
+`read_full` calls `read` repeatedly until the buffer has been filled, EOF has
+been reached, or an error other than `Interrupted` occurs. `read_exact` is
+similar to `read_full`, except that reaching EOF before filling the buffer is
+considered an error.
+
+# Motivation
+
+The `read` method may return fewer bytes than requested, and may fail with an
+`Interrupted` error if a signal is received during the call. This requires
+programs wishing to fill a buffer to call `read` repeatedly in a loop. This is
+a very common need, and it would be nice if this functionality were provided in
+the standard library. Many C and Rust programs have the same need, and solve it
+in the same way. For example, Git has [`read_in_full`][git], which behaves like
+the proposed `read_full`, and the Rust byteorder crate has
+[`read_full`][byteorder], which behaves like the proposed `read_exact`.
+[git]: https://github.com/git/git/blob/16da57c7c6c1fe92b32645202dd19657a89dd67d/wrapper.c#L246
+[byteorder]: https://github.com/BurntSushi/byteorder/blob/2358ace61332e59f596c9006e1344c97295fdf72/src/new.rs#L184
+
+# Detailed design
+
+The following methods will be added to the `Read` trait:
+
+``` rust
+fn read_full(&mut self, buf: &mut [u8]) -> Result<usize>;
```

Comment:
My opinion about this function is that it's too early to add, and if we do add it we don't want it in this form. While ignoring EINTR is a common desire, this method has a number of downsides:

* It's not composable with all existing uses of `T: Read`, everyone now needs to remember to call `read_full` instead of `read`.
* It's specific to just readers when writers also likely want the ability to "just ignore EINTR"
* This is a lossy method because if an error is encountered you have no way of knowing how much data was previously written.

If we were to add a function like this I believe we would add something akin to a `uninterrupted()` adapter. Like `take` this adapter would produce a new reader which would then ignore EINTR for *all operations*, solving the composability problem as well as "this should exist for writers" problems above. I don't think there's any way to solve the lossy problem.

Overall I think that this RFC is best fit by *only* proposing the `read_exact` method. It mirrors a common desire found in the community and it's what writers implement already.